### PR TITLE
Add in DB lazy load support

### PR DESCRIPTION
**Pull request by wburns**
Adds in required DB access that is configurable via application.properties.

Lazily loads the value from the DB if not present. It will cache null values. This change does not include expiration, which will be required for lazy cache only mode (eager does not require this).

Includes basis for weighed eviction where can estimate size based on JVM on heap. Still need to inspect JOL (Java Object Layout) to get these numbers more precise.

### CODE REVIEW

File: manager/src/main/java/io/gingersnapproject/database/DatabaseHandler.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+         throw new IllegalArgumentException("No rule found for " + rule);
+      }
+
+      // Have to do this until bug is fixed allowing injection of reactive Pool
+      PreparedQuery<RowSet<Row>> preparedQuery = pool.preparedQuery(ruleConfig.selectStatement());
+      var quarkusPreparedQuery = io.vertx.mutiny.sqlclient.PreparedQuery.<RowSet<Row>>newInstance(preparedQuery);
+
+//      return pool.preparedQuery(ruleConfig.selectStatement())
+      return quarkusPreparedQuery
+            .execute(Tuple.from(key.split(ruleConfig.plainSeparator())))
```
Comment (jabolina) [inline manager/src/main/java/io/gingersnapproject/database/DatabaseHandler.java]: JSON keys could raise a problem here. Or do only accept plain?

Comment (wburns) [reply manager/src/main/java/io/gingersnapproject/database/DatabaseHandler.java]: Haven't implemented JSON yet :D

File: manager/src/main/java/io/gingersnapproject/Caches.java [hunk 2]
```diff
@@ -5,40 +5,91 @@
...
+            })
+            .build(key -> {
+               Uni<String> dbUni = databaseHandler.select(name, key)
+                     // Make sure to use memoize, so that if multiple people subscribe to this it won't cause
+                     // multiple DB lookups
+                     .memoize().indefinitely();
+               // This will replace the pending Uni from the DB with a UniItem so we can properly size the entry
+               // Note due to how lazy subscribe works the entry won't be present in the map  yet
+               dbUni.subscribe()
+                     .with(result -> replace(name, key, dbUni, result));
```
Comment (wburns) [inline manager/src/main/java/io/gingersnapproject/Caches.java]: If an error happens we probably should remove the dbUni from the cache as well.

File: manager/src/main/java/io/gingersnapproject/util/UniItem.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package io.gingersnapproject.util;
```
Comment (ryanemerson) [inline manager/src/main/java/io/gingersnapproject/util/UniItem.java]: Nitpick, but is `.mutiny` a more meaningful package name?

File: manager/pom.xml [hunk 4]
```diff
@@ -30,6 +30,14 @@
             <artifactId>infinispan-extension</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
```
Comment (ryanemerson) [inline manager/pom.xml:42]: We should also add the service-binding extension:

```xml
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-kubernetes-service-binding</artifactId>
        </dependency>
```

File: manager/pom.xml [hunk 5]
```diff
@@ -30,6 +30,14 @@
...
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-reactive-mysql-client</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-reactive-oracle-client</artifactId>
+        </dependency>
```
Comment (ryanemerson) [inline manager/pom.xml:49]: I'm guessing this was included for your local testing at this stage?

As `quarkus.datasource.db-kind=` is required at build time, I guess there's no point ever having multiple DB dependencies simultaneously. Are you happy to remove this for now and we can introduce additional vendor dependencies when we have per DB builds?

File: manager/src/main/java/io/gingersnapproject/Caches.java [hunk 6]
```diff
@@ -5,40 +5,98 @@
...
+      // TODO: technically this says it removed something even if the Uni contained a null value prior
+      // We put a null uni into the map if a value existed before. This way we cache the tombstone.
+      return getOrCreateMap(name).asMap()
+            .replace(key, Uni.createFrom().nullItem()) != null;
    }
 
-   public ConcurrentMap<String, Cache<String, List<byte[]>>> getMultiMaps() {
-      return multiMaps;
+   // Returns an estimated size which includes tombstone values
+   public long estimatedSize(String name) {
```
Comment (ryanemerson) [inline manager/src/main/java/io/gingersnapproject/Caches.java]: There's no harm having this, but did you have a specific use-case in mind where we need to have this available via HotRod?

I assume we'll need it for /metrics.

File: manager/src/main/java/io/gingersnapproject/util/UniItem.java (hunk 3 above)
Comment (wburns) [reply manager/src/main/java/io/gingersnapproject/util/UniItem.java]: I was planning on putting various utility things here, but sure I can change it.

File: manager/pom.xml (hunk 4 above)
Comment (wburns) [reply manager/pom.xml:42]: I don't think this should be in this PR though right?

File: manager/pom.xml (hunk 5 above)
Comment (wburns) [reply manager/pom.xml:49]: I am more than fine removing until we have a PR DB build

File: manager/src/main/java/io/gingersnapproject/Caches.java (hunk 6 above)
Comment (wburns) [reply manager/src/main/java/io/gingersnapproject/Caches.java]: I can remove, sure. It was just a simple command that I could implement.

File: manager/pom.xml (hunk 4 above)
Comment (ryanemerson) [reply manager/pom.xml:42]: As we're using the reactive client I don't think any additional implementation is required, just the dependency, so I think it makes sense to include it now.

Comment (wburns) [reply manager/pom.xml:42]: Alright, as long as it is just a simple dep update :)

**Comment by wburns** [comment]
Updated

**Comment by ryanemerson** [comment]
Thanks @wburns

**Pull request merged**